### PR TITLE
Handle null partitions in P2P shuffling

### LINKED ISSUE #7837: p2p shuffling failed if Arrow infers different dtypes for partitions
**Describe the issue**:

The first partition is inferred as null while the second is inferred as string, this clashes when concatenating them

**Minimal Complete Verifiable Example**:

```python
from distributed import Client
client = Client()

data = [
    {'companies': [], 'id': 'a', "x": None},
    {'companies': [{'id': 3},  {'id': 5}], 'id': 'b', "x": None},
    {'companies': [{'id': 3}, {'id': 4}, {'id': 5}], 'id': 'c', "x": "b"},
    {'companies': [{'id': 9}], 'id': 'a', "x": "a"}
]
df = pd.DataFrame(data)
ddf = dd.from_pandas(df, npartitions=2)
ddf = ddf.shuffle(on='id', shuffle="p2p", ignore_index=True)
```

```
2023-05-15 16:08:50,846 - distributed.worker - WARNING - Compute Failed
Key:       ('shuffle-p2p-38d3d79031f25db9add57999b152c9b2', 1)
Function:  shuffle_unpack
args:      ('9d0c74e7ff3dcaba60ce1957cc329424', 1, 0)
kwargs:    {}
Exception: "RuntimeError('shuffle_unpack failed during shuffle 9d0c74e7ff3dcaba60ce1957cc329424')"

Traceback (most recent call last):
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_shuffle.py", line 77, in shuffle_unpack
    return _get_worker_extension().get_output_partition(
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_worker_extension.py", line 903, in get_output_partition
    return sync(self.worker.loop, shuffle.get_output_partition, output_partition)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/utils.py", line 418, in sync
    raise exc.with_traceback(tb)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/utils.py", line 391, in f
    result = yield future
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/tornado/gen.py", line 767, in run
    value = future.result()
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_worker_extension.py", line 529, in get_output_partition
    out = await self.offload(_)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_worker_extension.py", line 132, in offload
    return await asyncio.get_running_loop().run_in_executor(
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_worker_extension.py", line 526, in _
    df = convert_partition(data)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_arrow.py", line 57, in convert_partition
    return pa.concat_tables(shards)
  File "pyarrow/table.pxi", line 5139, in pyarrow.lib.concat_tables
    c_result_table = GetResultValue(
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
    raise ArrowInvalid(message)
pyarrow.lib.ArrowInvalid: Schema at index 1 was different: 
companies: list<item: struct<id: int64>>
id: string
x: null
_partitions: uint64
__index_level_0__: int64
vs
companies: list<item: struct<id: int64>>
id: string
x: string
_partitions: uint64
__index_level_0__: int64

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/patrick/Library/Application Support/JetBrains/PyCharm2023.1/scratches/scratch.py", line 521, in <module>
    ddf.compute()
  File "/Users/patrick/PycharmProjects/dask_dev/dask/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/Users/patrick/PycharmProjects/dask_dev/dask/dask/base.py", line 599, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/client.py", line 3224, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/client.py", line 2359, in gather
    return self.sync(
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/utils.py", line 351, in sync
    return sync(
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/utils.py", line 418, in sync
    raise exc.with_traceback(tb)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/utils.py", line 391, in f
    result = yield future
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/tornado/gen.py", line 767, in run
    value = future.result()
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/client.py", line 2222, in _gather
    raise exception.with_traceback(traceback)
  File "/Users/patrick/mambaforge/envs/dask-dev/lib/python3.10/site-packages/distributed/shuffle/_shuffle.py", line 81, in shuffle_unpack
    raise RuntimeError(f"shuffle_unpack failed during shuffle {id}") from e
RuntimeError: shuffle_unpack failed during shuffle 9d0c74e7ff3dcaba60ce1957cc329424
```
**Anything else we need to know?**:

Found here: https://dask.discourse.group/t/how-to-drop-duplicates-by-string-id-for-a-large-dataframe/1835

cc @hendrikmakait 

**Environment**:

- Dask version: main
- Python version: 3.10
- Operating System: mac
- Install method (conda, pip, source):

### PR DESCRIPTION
**Pull request by detroyejr**
Ran into this issue after upgrading to 2023.6. Luckily, it's an easy fix.

Columns with sparse data can have partitions with empty/null columns. This creates a conflict of data types between the partitions and an error during concatenation.

Closes #7837

- [ x] Tests added / passed
- [ x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

Admins can comment `ok to test` to allow this one PR to run or `add to allowlist` to allow all future PRs from the same author to run.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @detroyejr! 

@hendrikmakait @phofl do either of you have bandwidth to review?

**Review by jrbourbeau — COMMENTED** [review]
Also, could you add a small test for this? Something similar to the snippet in the original issue should be sufficient

**Comment by detroyejr** [comment]
Test added. Seems to do the trick, but let me know if you have any feedback.

File: distributed/tests/test_client.py
Comment (phofl) [inline distributed/tests/test_client.py:3216]: You can use 

```
pytest.importorskip("pyarrow")
```

here. That's a bit shorter and less complicated.

**Review by phofl — COMMENTED** [review]
Test generally looks good to me and should cover the bug. 

Needs another pair of eyes from someone who is more familiar with distributed tests, @hendrikmakait maybe?

File: distributed/tests/test_client.py
Comment (hendrikmakait) [inline distributed/tests/test_client.py:3215]: Thanks for adding a test case for the original issue. We currently group all tests related to P2P-shuffling in `distributed/shuffle/tests/test_shuffle.py` for discovery, please move it there. I'd also suggest you take a look at `test_concurrent`. It should serve as a good boilerplate and might help you simplify your test case (for example, skipping tests if packages are not installed is already taken care of): https://github.com/dask/distributed/blob/c520f1cf652882dc15c7cb8faf456ce5dc26f478/distributed/shuffle/tests/test_shuffle.py#L180

To make sure everything works correctly, we should also test that the resulting dataframe matches the expected pandas dataframe. You can use `dask.dataframe.utils.assert_eq` for this.

**Review by hendrikmakait — CHANGES_REQUESTED** [review]
Thanks, @detroyejr, for fixing this! The code looks good to me. I would like some improvements for the test case before approving this.

File: distributed/shuffle/tests/test_shuffle.py
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:1843]: The `assert` is not necessary, `assert_eq` already takes care of that and throws a meaningful `AssertionError`. I'd also suggest making use of the file-level import of `dask.dataframe` to make the assertion more readable.   

```suggestion
    dd.assert_eq(result, df)
```

**Review by hendrikmakait — COMMENTED** [review]
@detroyejr: Overall, the test looks great now, I only have one minor suggestion for improvement.

**Review by hendrikmakait — APPROVED** [review]
Thanks, @detroyejr, great work! I've merged the latest `main` to avoid CI failures caused by upstream changes and will merge once CI is done and green-ish.

**Comment by detroyejr** [comment]
Awesome, thanks for all your help!

**Pull request merged**
